cmd/syncthing/gui.go

```diff
@@ -1567,15 +1580,31 @@ func browseFiles(current string, fsType fs.FilesystemType) []string {
 
 	fs := fs.NewFilesystem(fsType, searchDir)
 
-	subdirectories, _ := fs.Glob(searchFile + "*")
+	subdirectories, _ := fs.Glob(makeSearchGlobPattern(searchFile))
 
 	ret := make([]string, 0, len(subdirectories))
+	matchcount := make([]int, 0, len(subdirectories))
 	for _, subdirectory := range subdirectories {
 		info, err := fs.Stat(subdirectory)
 		if err == nil && info.IsDir() {
+			idx := len(ret)
 			ret = append(ret, filepath.Join(searchDir, subdirectory)+pathSeparator)
+			matchcount = append(matchcount, 0);
+			for i := 0; i < len(subdirectory) && i < len(searchFile); i += 1 {
+				if subdirectory[i] == searchFile[i] {
+					matchcount[idx] += 1
+				}
+			}
 		}
 	}
+	// sort by number of matches with pattern, largest number of matches first
+	// ties are resolved by alphabetical order
+	sort.Slice(ret, func(i, j int) bool {
+		if matchcount[i] == matchcount[j] {
+			return strings.ToLower(ret[i]) < strings.ToLower(ret[j])
+		}
+		return matchcount[i] > matchcount[j]
+	})
```

Comment:
It's strange that ret is sorted based on some other array. Is it not easier to use a map or a slice of structs where it has the match and the count?

Reply:
hmm, with a map or a slice of structs I then need allocate another array at the end for the return value. Is this preferred?